truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/source/SourceSection.java

```diff
@@ -76,6 +83,30 @@
         this.startColumn = startColumn;
         this.charIndex = charIndex;
         this.charLength = charLength;
+        this.tags = tags;
+        assert tagsAreNonNullAndInterned(tags) : "All tags set for a source section must be interned and non-null.";
+    }
+
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    private static boolean tagsAreNonNullAndInterned(String[] tags) {
+        for (int i = 0; i < tags.length; i++) {
+            if (tags[i] == null) {
+                return false;
+            }
+            if (tags[i].intern() != tags[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns a set of tags that are associated with this source section as array. The returned
+     * array must be considered read-only. All elements of the array were asserted to be non-null
+     * and interned for faster comparison.
+     */
+    public String[] getTags() {
```

Comment:
You are using this method for JSON serialization(!) - that is exactly my point: This method has never been designed to help you with serialization, so from the point of view of all the use-cases we have, you are *misusing it*.

Don't get me wrong. You have right to *misuse it* as it is there and you can use it anyway you like. But then it has to stay there up until the end of times (those are the rules I want us to follow). And honestly, it is a bit too strong commitment for a method that has only been introduced for lack of ability to negotiate better solution.

But it is not my decision, I am just warning you guys that I will block attempts to remove the method in the future, if it is introduced now.